Prototope/Behaviors.swift

```diff
@@ -0,0 +1,194 @@
+//
+//  Collisions.swift
+//  Prototope
+//
+//  Created by Saniul Ahmed on 07/02/2015.
+//  Copyright (c) 2015 Khan Academy. All rights reserved.
+//
+
+import Foundation
+
+//MARK: Public Behavior Types
+
+// Protocol for types describing the "configuration" of a behavior.
+public protocol BehaviorType {}
+
+// Value type describing the configuration of a collision behavior, specifying
+// the kind of the interaction (the collision conditions which should trigger the handler)
+// the layer with which the host layer is supposed to be colliding and a handler function
+public struct CollisionBehavior: BehaviorType {
+    public enum Kind {
+        case Entering
+        case Leaving
+    }
+    
+    let kind: Kind
+    let otherLayer: Layer
+    let handler: () -> Void
+    
+    public init(on kind: Kind, _ otherLayer: Layer, handler: ()->Void) {
+        self.kind = kind
+        self.otherLayer = otherLayer
+        self.handler = handler
+    }
+}
+
+//MARK: Behavior Bindings
+
+// Abstract. Describes a relationship between a layer and a single instance of a behavior.
+// Subclasses should encapsulate the state necessary to handle the behavior correctly.
+class BehaviorBinding: Equatable, Hashable {
```

Comment:
Mmm, that was what I wanted to do initially, but:
`Equatable` uses `Self`, which makes the declaration `var registeredBindings: Set<BehaviorBinding>` in `BehaviorStore` impossible – “Protocol ‘BehaviorBinding’ can only be used as a generic constraint...”. This would force BehaviorStore to only work with a single type of behaviors, which is obviously not what we want.

If there’s a nicer way out of this other than an abstract class I would be very glad to learn it, this isn’t the first time I encounter this limitation.